File: geo3x3.fs
namespace Geo3x3

module Geo3x3 =
    let encode lat lng level =
        if level < 1 then
            ""
        else
            let (res, lng1) =
                if lng >= 0.0 then
                    ("E", lng)
                else
                    ("W", lng + 180.0) in
            let lng2 = lng1 in
            let lat2 = (90.0 - lat) in (* 0:the North Pole,  180:the South Pole *)
            let rec loop i lat lng unit =
                if i >= level then
                    ""
                else
                    let x = int (lng / unit) in
                    let y = int (lat / unit) in
                    let n = x + y * 3 + 1 in
                    let lng2 = lng - (float x) * unit in
                    let lat2 = lat - (float y) * unit in
                    (string n) + (loop (i + 1) lat2 lng2 (unit / 3.0)) in
            res + (loop 1 lat2 lng2 (180.0 / 3.0))

    let decode code = (* (35., 135., 14, 0.1) *)
        let clen = String.length code in
        if clen < 1 then
            (0.0, 0.0, 0, 180.0) (* err *)
        else
            let c = code.[0] in
            let (flg, beg) =
                if c = '-' || c = 'W' then
                    (true, 1)
                else if c = '+' || c = 'E' then
                    (false, 1)
                else
                    (false, 0) in
            let rec loop i lat lng level unit =
                if i >= clen then
                    (lat, lng, level, unit)
                else
                    let unit2 = unit / 3.0 in
                    let n = (int code.[i]) - 49 in (* ignore err check *)
                    let lat2 = lat + (float (n / 3)) * unit2 in
                    let lng2 = lng + (float (n % 3)) * unit2 in
                    loop (i + 1) lat2 lng2 (level + 1) unit2 in
            let (lat, lng, level, unit) = loop beg 0.0 0.0 1 180.0 in
            let lat2 = lat + (unit / 2.0) in
            let lng2 = lng + (unit / 2.0) in
            let lat3 = 90.0 - lat2 in
            let lng3 = if flg then lng2 - 180.0 else lng2 in
            (lat3, lng3, level, unit)

File: geo3x3.erl
-module(geo3x3).
-export([encode/3, decode/1]).

encode_loop(Level, I, Lat, Lng, Unit) ->
    if
        I >= Level -> [];
        true ->
            X = trunc(Lng / Unit),
            Y = trunc(Lat / Unit),
            N = X + Y * 3 + 1,
            Lng2 = Lng - float(X) * Unit,
            Lat2 = Lat - float(Y) * Unit,
            [ N + 48 | encode_loop(Level, I + 1, Lat2, Lng2, Unit / 3.0)]
    end.

encode(Lat, Lng, Level) ->
    if
        Level < 1 -> "";
        true -> 
            {Res, Lng1} =
                if
                    Lng >= 0.0 -> {"E", Lng};
                    true -> {"W", Lng + 180.0}
                end,
            Lng2 = Lng1,
            Lat2 = (Lat + 90.0), % 0:the North Pole,  180:the South Pole
            [Res | encode_loop(Level, 1, Lat2, Lng2, 180.0 / 3.0)]
    end.

decode_loop(Code, Clen, I, Lat, Lng, Level, Unit) ->
    if
        I >= Clen -> {Lat, Lng, Level, Unit};
        true ->
            Unit2 = Unit / 3.0,
            [N | _] = string:substr(Code, I + 1, 1),
            N2 = N - 49, % ignore err check
            Lat2 = Lat + (N2 div 3) * Unit2,
            Lng2 = Lng + (N2 rem 3) * Unit2,
            decode_loop(Code, Clen, I + 1, Lat2, Lng2, Level + 1, Unit2)
    end.

decode(Code) ->
    Clen = length(Code),
    if
        Clen < 1 -> {0.0, 0.0, 0, 180.0}; % err
        true ->
            [C | _] = Code,
            {Flg, Beg} =
                if
                    (C == $-) or (C == $W) -> {true, 1};
                    (C == $+) or (C == $E) -> {false, 1};
                    true -> {false, 0}
                end,
            {Lat, Lng, Level, Unit} = decode_loop(Code, Clen, Beg, 0.0, 0.0, 1, 180.0),
            Lat2 = Lat + (Unit / 2.0),
            Lng2 = Lng + (Unit / 2.0),
            Lat3 = Lat2 - 90.0,
            Lng3 = if
                    Flg -> Lng2 - 180.0;
                    true -> Lng2
                end,
            {Lat3, Lng3, Level, Unit}
    end.
